README
This is the README file for the Perl module Device::TNC

This module was created created purely because I knew Perl better than C and
Perl has so many lovely modules available to make a developer life easy.

What does Device::TNC provide?

The module Device::TNC provides the basis of a generic set of modules that can
be used to interface with a Terminal Node Controller (TNC).

What's a TNC?

If you don't know then this module probably won't interest you.
A TNC provides an interface between radio communication and the user.
In some cases this is via a display on the TNC but in other cases this is simply
a connection to a serial port on a computer.
The serial port could be an RS232 connection, a USB to Serial adaptor or a USB
conneciton with a software defined serial port.

The initial versions of this module have been developed on Linux and Windows 
2000 with a KISS mode TNC (TNC-X).
It is hoped (but not guarenteed) that this module will have been developed in
such a way as it's easily extended for use with other TNC's.
I expect the API to change as people subclass this module to work with other
TNC's.
If you do this sort of thing please let me know so that I can assist with your
development work and any re-design that may (probably will) be necessary.

Installation

Installation follows the usual Perl module installation procedure.

perl Makefile.PL
make
make test (This is optional but recommended)
make install

Copyright (C) 2007, Australian Space Research Institute.

This module is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.



Changes
Release 0.02
  * Fixed a bug that caused most of the data in the INFO field to be ignored.
  * Now the full INFO field for the captured frame is being included.
  * The Device::TNC::KISS module can now read frames from a raw KISS log file.
  * The FCS bits at the end of AX.25 frames are not present in the frames I
  collect using my TNC-X. Thus I cannot parse this data and return it in the
  structure returned by read_ax25_frame for now the field has been removed.

Release 0.01
  * Initial creation.
  * Has been tested with a TNC-X under Linux (2.6.16 kernel) and Windows 2000.
  * There is a bug on Windows where the KISS TNC doesn't open properly for some
  reason. To work around this I've found that if you connect to the TNC using
  UI-View and the quite UI-View the TNC can be used after that. I have not been
  able to track down the cause of this issue however.


lib/Device/TNC/KISS.pm

=head1 NAME

Device::TNC::KISS - A module to talk to a KISS mode TNC

=head1 DESCRIPTION

This module trys to implement an easy way to talk to a KISS mode Terminal Node
Controller (TNC) such as the TNC-X via a serial port.

=head1 SYNOPSIS

  use Device::TNC::KISS;
  my %tnc_config = (
    'port' => ($Config{'osname'} eq "MSWin32") ? "COM3" : "/dev/TNC-X",
    'baudrate' => 9600,
    'warn_malformed_kiss' => 1,
    'raw_log' => "raw_packet.log",
  );
  my $kiss_tnc = new Device::TNC::KISS(%tnc_config);

  my $kiss_frame = $kiss_tnc->read_kiss_frame();
  my @kiss_frame = $kiss_tnc->read_kiss_frame();

  my ($kiss_type, $hdlc_frame) = $kiss_tnc->read_hdlc_frame();
  my ($kiss_type, @hdlc_frame) = $kiss_tnc->read_hdlc_frame();

This module was developed on Linux and Windows. It should work for any UNIX like
operating system where the Device::SerialPort module can be used.

=cut

package Device::TNC::KISS;

####################
# Standard Modules
use strict;
use Config;
use FileHandle 2.0;
# Custom modules
use Device::TNC;

####################
# Variables
use vars qw(@ISA @EXPORT @EXPORT_OK $VERSION);
@ISA = qw( Device::TNC );
@EXPORT = qw();
@EXPORT_OK = qw();
$VERSION = 0.01;
$| = 1;


####################
# Functions

if ($Config{'osname'} eq "MSWin32")
{
	require Win32::SerialPort;
	Win32::SerialPort->import( qw( :STAT 0.19 ) );
}
else
{
	require Device::SerialPort;
}

my $FEND  = 0xC0; # Frame end
my $FESC  = 0xDB; # Frame escape
my $TFEND = 0xDC; # Transposed frame end
my $TFESC = 0xDD; # Transposed frame escape

my $m_port = undef;
my $m_port_name = undef;
my $m_raw_log = undef;

################################################################################

=head2 B<new()>

 my %port_data = { 'option' => 'value' };
 my $kiss_tnc = new Device::TNC::KISS(%port_data);

The new method creates and returns a new Device::TNC::KISS object that can be
used to communicate with a KISS mode terminal node controller.

The method requires that a hash of settings be passed.
If the port and baudrate are not set in the passed settings or the port cannot
be opened an error message is printed and undef returned.

=head3 Options and values

=over 4

=item B<port>

This sets the serial port name as appropriate for the operating system.
For UNIX this will be something like /dev/ttyS0 and for Windows this will be
something like COM1

=item B<baudrate>

The baud rate is the speed the TNC is configured to talk at. i.e. 1200 baud.

=item B<warn_malformed_kiss>

To see warnings about malformed KISS frames then set this option to a true value

=item B<raw_log>

If you want to keep a log of the raw packets that are read then set raw_log to
the name of the log file you which to write to. If there is an error opening the
log file the

=back

The returned object contains a reference to a serial port object which is either a
Device::SerialPort (UNIX) or Win32::SerialPort (Windows) object.

The serial port is initialised will the following

 $kiss_tnc->{'PORT'}->parity("none");
 $kiss_tnc->{'PORT'}->databits(8);
 $kiss_tnc->{'PORT'}->stopbits(1);
 $kiss_tnc->{'PORT'}->handshake("none");
 $kiss_tnc->{'PORT'}->read_interval(100) if $Config{'osname'} eq "MSWin32";
 $kiss_tnc->{'PORT'}->read_char_time(0);
 $kiss_tnc->{'PORT'}->read_const_time(1000);

For more details on this see the documentation for Device::SerialPort (UNIX) or
Win32::SerialPort (Windows).

=cut

sub new
{
	my $class = shift;
	my %port_data = @_;

	my $baudrate = undef;
	my %data;
	foreach my $key (keys %port_data)
	{
		if (lc($key) eq "port")
		{
			$m_port_name = $port_data{$key};
		}
		if (lc($key) eq "baudrate")
		{
			$baudrate = $port_data{$key};
		}
		if (lc($key) eq "warn_malformed_kiss")
		{
			$data{'WARN_MALFORMED_KISS'} = 1;
		}
		if (lc($key) eq "raw_log")
		{
			$m_raw_log = new FileHandle();
			if ($m_raw_log->open(">>$port_data{$key}"))
			{
				$m_raw_log->autoflush(1);
				$data{'RAW_LOG'} = $m_raw_log;
			}
			else
			{
				warn "Warning: Cannot open raw log file \"$port_data{$key}\" for append: $!\n";
			}
		}
	}
	unless ($m_port_name)
	{
		warn "Error: No port was specified in the past data.\n";
		return undef;
	}
	unless ($baudrate)
	{
		warn "Error: No baudrate was specified in the past data.\n";
		return undef;
	}

	if ($Config{'osname'} eq "MSWin32")
	{
		$m_port = new Win32::SerialPort($m_port_name) or
			warn "Error: Cannot open serial port \"$m_port_name\": $^E\n";
	}
	else
	{
		$m_port = new Device::SerialPort($m_port_name) or
			warn "Error: Cannot open serial port \"$m_port_name\": $!\n";
	}
	$m_port->baudrate($baudrate);
	$m_port->parity("none");
	$m_port->databits(8);
	$m_port->stopbits(1);
	$m_port->handshake("none");
	$m_port->read_interval(100) if $Config{'osname'} eq "MSWin32";
	$m_port->read_char_time(0);
	$m_port->read_const_time(1000);

	$data{'PORT'} = $m_port;
	my $self = bless \%data, $class;

	return $self;
}

# When we close try and close the port too.
DESTROY
{
	# Close the serial port
	$m_port->close() or
		warn "Error: Failed to the serial port \"$m_port_name\": $!\n";

	# Close the raw log file if we have opened it.
	$m_raw_log->close() if $m_raw_log;

}

################################################################################

=head2 B<read_kiss_frame()>

 my $kiss_frame = $kiss_tnc->read_kiss_frame();
 my @kiss_frame = $kiss_tnc->read_kiss_frame();

This method reads a KISS frame from the TNC and returns it. This has not had
the FEND, FESC, TFEND and TFESC bytes stripped out.

=cut

sub read_kiss_frame
{
	my $self = shift;
	my @frame;
	my $fend_count = 0;
	while(1)
	{
		# Processing one byte at a time makes things nice and easy
		my ($count,$saw) = $self->{'PORT'}->read(1);
		if ($count > 0)
		{
			$self->{'RAW_LOG'}->write($saw) if $self->{'RAW_LOG'};
			$fend_count++ if (ord($saw) == $FEND);
			# Make sure we don't add bytes to the frame that are before the first FEND
			push @frame, $saw if $fend_count > 0;

			if ($fend_count == 2)
			{
				if (scalar @frame > 2)
				{
					# We have data in the frame so return it.
					last;
				}
				else
				{
					# we have an empty frame or we got the end of one frame and the start of another.
					# So start the search again.
					$fend_count--;
					@frame = ($saw);
				}
			}
		}
	}
	$self->{'LAST_KISS_FRAME_LENGTH'} = scalar @frame;

	if (wantarray)
	{
		return @frame;
	}
	else
	{
		my $frame = join '', @frame;
		return $frame;
	}
}

################################################################################

=head2 B<read_hdlc_frame()>

 my ($kiss_type, $hdlc_frame) = $kiss_tnc->read_hdlc_frame();
 my ($kiss_type, @hdlc_frame) = $kiss_tnc->read_hdlc_frame();

This method reads a KISS frame from the TNC and strips out the KISS FEND, FESC,
TFEND and TFESC bytes and returns the KISS type byte followed by the HDLC frame.

The value of type indicator byte is use to distinguish between command and data
frames.

From: The KISS TNC: A simple Host-to-TNC communications protocol
L<http://people.qualcomm.com/karn/papers/kiss.html>

To distinguish between command and data frames on the host/TNC link, the first
byte of each asynchronous frame between host and TNC is a "type" indicator. This
type indicator byte is broken into two 4-bit nibbles so that the low-order
nibble indicates the command number (given in the table below) and the
high-order nibble indicates the port number for that particular command.
In systems with only one HDLC port, it is by definition Port 0. In multi-port
TNCs, the upper 4 bits of the type indicator byte can specify one of up to
sixteen ports. The following commands are defined in frames to the TNC (the
"Command" field is in hexadecimal):

 Command       Function         Comments
   0           Data frame       The  rest  of the frame is data to
                                be sent on the HDLC channel.

   1           TXDELAY          The next  byte  is  the  transmitter
                                keyup  delay  in  10 ms units.
                		The default start-up value is 50
                                (i.e., 500 ms).

   2           P                The next byte  is  the  persistence
                                parameter,  p, scaled to the range
                                0 - 255 with the following
                                formula:

                                         P = p * 256 - 1

                                The  default  value  is  P  =  63
                                (i.e.,  p  =  0.25).

   3           SlotTime         The next byte is the slot interval
                                in 10 ms units.
                                The default is 10 (i.e., 100ms).

   4           TXtail           The next byte is the time to hold
                                up the TX after the FCS has been
                                sent, in 10 ms units.  This command
                                is obsolete, and is included  here
                                only for  compatibility  with  some
                                existing  implementations.

   5          FullDuplex        The next byte is 0 for half duplex,
                                nonzero  for full  duplex.
                                The  default  is  0
                                (i.e.,  half  duplex).

   6          SetHardware       Specific for each TNC.  In the
                                TNC-1, this command  sets  the
                                modem speed.  Other implementations
                                may use this function  for   other
                                hardware-specific   functions.

   FF         Return            Exit KISS and return control to a
                                higher-level program. This is useful
                                only when KISS is  incorporated
                                into  the TNC along with other
                                applications.

The following types are defined in frames to the host:

Type			Function		Comments

  0                 Data frame       Rest of frame is data from
                                     the HDLC channel.

No other types are defined; in particular, there is no provision for
acknowledging data or command frames sent to the TNC. KISS implementations must
ignore any unsupported command types. All KISS implementations must implement
commands 0,1,2,3 and 5; the others are optional.

=cut

sub read_hdlc_frame
{
	my $self = shift;

	# This is what we will return
	my @frame;
	my $type;

	# This is not the simplest method of getting the data but it allows for
	# for catching errors.
	my @kiss = $self->read_kiss_frame();
	for (my $location = 0; $location <= $#kiss; $location++)
	{
		if ($location == 0)
		{
			# We should find an FEND here. If not print a warning.
			unless (ord($kiss[$location]) == $FEND)
			{
				warn "Warning: Malformed KISS frame read. Didn't start with FEND\n" if $self->{'WARN_MALFORMED_KISS'};
			}
		}
		elsif ($location == $#kiss)
		{
			# We should find an FEND here. If not print a warning.
			unless (ord($kiss[$location]) == $FEND)
			{
				warn "Warning: Malformed KISS frame read. Didn't end with FEND\n" if $self->{'WARN_MALFORMED_KISS'};
			}
		}
		elsif ($location == 1)
		{
			# This is the type byte
			$type = $kiss[$location];
		}
		else
		{
			# this is the data but may contains transposed bytes
			# FEND, FESC, TFEND and TFESC
			if (ord($kiss[$location]) == $FESC)
			{
				# Entered frame escape mode
				if (ord($kiss[$location + 1]) == $TFESC)
				{
					#warn "Un-Transposed a FESC\n";
					push @frame, pack("c", $FESC);
					$location++;
				}
				elsif (ord($kiss[$location + 1]) == $TFEND)
				{
					#warn "Un-Transposed a FEND\n";
					push @frame, pack("c", $FEND);
					$location++;
				}
				elsif (ord($kiss[$location + 1]) != $TFESC)
				{
					warn "Warning: Malformed KISS frame read. Expected TFESC after FESC\n" if $self->{'WARN_MALFORMED_KISS'};
				}
				elsif (ord($kiss[$location + 1]) != $TFEND)
				{
					warn "Warning: Malformed KISS frame read. Expected TFEND after FESC\n" if $self->{'WARN_MALFORMED_KISS'};
				}
			}
			else
			{
				push @frame, $kiss[$location];
			}
		}
	}

	if (wantarray)
	{
		return $type, @frame;
	}
	else
	{
		my $frame = join '', @frame;
		return $type, $frame;
	}
}

1;

__END__

=head1 SEE ALSO

 Device::SerialPort
 Win32::SerialPort

L<http://people.qualcomm.com/karn/papers/kiss.html>

=head1 AUTHOR

R Bernard Davison E<lt>bdavison@asri.org.auE<gt>

=head1 COPYRIGHT

Copyright (C) 2007, Australian Space Research Institute.

This module is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut


lib/Device/TNC.pm

=head1 NAME

Device::TNC - A generic interface to a TNC

=head1 DESCRIPTION

This module implements a generic interface to a Terminal Node Controller (TNC).

It loads sub classes that provide the low level interface for the appropriate
TNC to be used and provides higher level methods to return frames of data to the
user is human readable form.

=head1 SYNOPSIS

  use Device::TNC;
  my $tnc_type = 'KISS';
  my %tnc_config = (
    'port' => ($Config{'osname'} eq "MSWin32") ? "COM3" : "/dev/TNC-X",
    'baudrate' => 9600,
    'warn_malformed_kiss' => 1,
    'raw_log' => "raw_packet.log",
  );
  my $tnc = new Device::TNC($tnc_type, %tnc_config);
  die "Error: Something went wrong connecting to the TNC.\n" unless $tnc;

  while (1)
  {
    my $data = $tnc->read_ax25_frame();
    my $repeaters = join ", ", @{$data->{'ADDRESS'}->{'REPEATERS'}};
    my $info = join "", @{$data->{'INFO'}};
    print "From: $data->{'ADDRESS'}->{'SOURCE'} ";
    print "To: $data->{'ADDRESS'}->{'DESTINATION'} ";
    print "via $repeaters\n";
    print "Data: $info\n";
  }

=cut

package Device::TNC;

####################
# Standard Modules
use strict;
use Config;
use Data::Translate;
# Custom modules

####################
# Variables
use vars qw(@ISA @EXPORT @EXPORT_OK $VERSION);
@ISA = qw();
@EXPORT = qw();
@EXPORT_OK = qw();
$VERSION = 0.02;
$| = 1;

my $translator = new Data::Translate();

# I'm not sure what this was for... It came from the KISS2ASC program that I
# originally based a lot of this on. I've since updated everything that is here
# to match/comply with what the AX.25 standard doc says.
# I'm not using it now and will remove soon unless I find the reason it was in
# KISS2ASC.

# pollcode gives the flags used to represent the various sets of poll and final flags.
# It is indexed by three bits:
# pollcode[0x10 bit of control][0x80 bit of destcall][0x80 bit of fromcall]
#
# We use the following codes, Followed WA8DED:
#       ! Version 1 with poll/final
#       ^ Version 2 command without poll
#       + Version 2 command with poll
#       - Version 2 response with final
#       v Version 2 response without final
# 0 in the table indicates no charater is to be output.
#
my @m_pollcode;
$m_pollcode[0]->[0]->[0] = 0;
$m_pollcode[0]->[0]->[1] = 'v';
$m_pollcode[0]->[1]->[0] = '^';
$m_pollcode[0]->[1]->[1] = 0;
$m_pollcode[1]->[0]->[0] = '!';
$m_pollcode[1]->[0]->[1] = '-';
$m_pollcode[1]->[1]->[0] = '+';
$m_pollcode[1]->[1]->[1] = '!';

# This is how this array is used.
#my $val1 = ($control & 0x10) ? 1 : 0;
#my $val2 = (ord($frame[6]) & 0x80) ? 1 : 0;
#my $val3 = (ord($frame[13]) & 0x80) ? 1 : 0;
#$data{'POLLCHAR'} = $m_pollcode[$val1]->[$val2]->[$val3];

####################
# Functions

################################################################################

=head2 B<new()>

 my $type = "KISS";
 my %tnc_data = { 'option' => 'value' };
 my $tnc = new Device::TNC($type, %tnc_data);

The new method creates and returns a new Device::TNC object that can be
used to communicate with a Terminal Node Controller (TNC) of the type passed.

The method requires that the first passed argument be the type of TNC to connect
to. This will try and load the appropriate module for the TNC type.

The subsequent options are passed to the module that is loaded to connect to the
desired TNC.

For more details on these options see the module documentation for the TNC type.

=cut

sub new
{
	my $class = shift;
	my $type = uc(shift);
	my %tnc_data = @_;
	my $tnc = undef;

	unless (scalar $type)
	{
		warn "Error: No TNC type passed.\n";
		return undef;
	}

	my $load_module = "require Device::TNC::$type;\n";
	$load_module .= '$tnc' . " = new Device::TNC::$type(%tnc_data);\n";
	eval $load_module;
	if ($@)
	{
		warn "$@\n";
		warn "Error: Failed to load Device::TNC::$type\n";
	}

	return $tnc;
}

################################################################################

=head2 B<read_ax25_frame()>

 my $ax25_data = $tnc->read_ax25_frame();
 my %ax25_data = $tnc->read_ax25_frame();

This method reads a HDLC frame from the TNC and returns a structure as either a
hash or a hash reference that contains the fields of the AX.25 frame.

The structure of the returned data is like the following.

  {
    'INFO' => [
      '/', '0', '6', '4', '6', '5', '8', 'h', '3', '3', '5', '0', '.', '0', '0',
      'S', '\\', '1', '5', '1', '1', '2', '.', '0', '0', 'E', 'O', '2', '2', '6',
      '/', '0', '0', '0', '/', 'A', '=', '0', '0', '0', '1', '1', '1'
    ],
    'PID' => 'F0',
    'CONTROL' => {
      'POLL_FINAL' => 0,
      'FIELD_TYPE' => 'UI',
      'FRAME_TYPE' => 'U'
    },
    'ADDRESS' => {
      'DESTINATION' => 'APT311',
      'REPEATERS' => [
        'WIDE1-1',
        'WIDE2-2'
        ],
      'SOURCE' => 'VK2KFJ-7'
    }
  }

While developing this module I only received U (UI) type frames and so
development of the code to work with I and S frames didn't really progress.
If anyone want's to read I or S frames please let me know and I'll have a look
at implementing them. Please create a KISS log of the data and email it to me.

According to the AX.25 standard there are supposed to be two bytes at the and of
the frame that contain the Frame Check Sequense (FCS) commonly know as a CRC.

While developing this code I have found that these two bytes are bot passed from
my TNC-X to the user. I'm not sure if this is how all TNC's work however.
Again if someone could create a raw KISS log of some packets that contain this
data and send it to me then I can have a look at implementing a check in this
module.

=cut

sub read_ax25_frame
{
	my $self = shift;
	my %data;
	my ($type, @frame) = $self->read_hdlc_frame();

	my $location = 0;
	# Get the destination
	for (my $loc = 0; $loc < 7; $loc++)
	{
		my $byte = ord($frame[$location]);
		if ($byte != 0x40)
		{
			my $shift_byte = ($byte >> 1);
			my ($s, $ascii) = $translator->h2a(sprintf("%X",$shift_byte));
			if ($loc == 6)
			{
				my $ssid = ($byte & 0x1E) >> 1;
				$data{'ADDRESS'}->{'DESTINATION'} .= "-$ssid" if $ssid > 0;
			}
			else
			{
				$data{'ADDRESS'}->{'DESTINATION'} .= $ascii;
			}
		}
		$location++;
	}
	# Get the source
	for (my $loc = 0; $loc < 7; $loc++)
	{
		my $byte = ord($frame[$location]);
		if ($byte != 0x40)
		{
			my $shift_byte = ($byte >> 1);
			my ($s, $ascii) = $translator->h2a(sprintf("%X",$shift_byte));
			if ($loc == 6)
			{
				my $ssid = ($byte & 0x1E) >> 1;
				$data{'ADDRESS'}->{'SOURCE'} .= "-$ssid" if $ssid > 0;
			}
			else
			{
				$data{'ADDRESS'}->{'SOURCE'} .= $ascii;
			}
		}
		$location++;
	}

	# Find the repeaters if any.
	@{$data{'ADDRESS'}->{'REPEATERS'}} = ();
	my $control = ord($frame[$location]);
	while (($control & 1) == 0)
	{
		my $repeater = "";
		for (my $loc = 0; $loc < 7; $loc++)
		{
			my $byte = ord($frame[$location]);
			if ($byte != 0x40)
			{
				my $shift_byte = ($byte >> 1);
				my ($s, $ascii) = $translator->h2a(sprintf("%X",$shift_byte));
				if ($loc == 6)
				{
					my $ssid = ($byte & 0x1E) >> 1;
					#printf(" SSID: %d\n", $ssid);
					$repeater .= "-$ssid" if $ssid > 0;
				}
				else
				{
					$repeater .= $ascii;
				}
			}
			$location++;
		}
		push @{$data{'ADDRESS'}->{'REPEATERS'}}, $repeater;
		$control = ord($frame[$location]);
	}

	# Now find the frame type
	$location++;
	if (($control & 1) == 0) # Information (I) frame found
	{
		# No data gathered to work with here.

		# Control field is 1 or 2 bytes
#		push @{$data{'CONTROL'}}, $control, ord($frame[$location]);
#		push @{$data{'CRAP'}->{'CONTROL_BIN'}}, $translator->d2b($control), $translator->d2b(ord($frame[$location]));

#		$data{'CRAP'}->{'FRAME_TYPE'} = sprintf("I%d%d", (($control & 0xE0) >> 5), (($control & 0x0E) >> 1));
#		$data{'PID'} = sprintf("%02X", (ord($frame[$location]) & 0xFF));
#		while ( (my $byte = $frame[$location]) and ( ($location ne $#frame) or ($location ne $#frame - 1)) )
#		{
#			#my ($s, $ascii) = $translator->h2a(sprintf("%X",$byte));
#			push @{$data{'INFO'}},  $byte;
#			$location++;
#		}
	}
	elsif (($control & 3) == 1) # Supervisory (S) frame found
	{
		# No data gathered to work with here.

		# Control field is 1 or 2 bytes
#		push @{$data{'CONTROL'}}, $control, ord($frame[$location]);
#		push @{$data{'CRAP'}->{'CONTROL_BIN'}}, $translator->d2b($control), $translator->d2b(ord($frame[$location]));

#		$data{'CRAP'}->{'FRAME_TYPE'} = sprintf("%s%d", (($control & 0x0C) >> 2), (($control & 0xE0) >> 5) );
#		$data{'CRAP'}->{'FRAME_TYPE'} .= sprintf(" FRAME TYPE IN HEX = %X for 0C or maybe %X for EF it's number is %d", ($control & 0x0C), ($control & 0xEF), (($control & 0xE0) >> 5) );
	}
	elsif (($control & 0xEF) == 0x03) # Unnumbered (U) frame found
	{
		# Control field is 1 byte
		$data{'CONTROL'}->{'FRAME_TYPE'} = "U";
		$data{'CONTROL'}->{'POLL_FINAL'} = ($control & 0x10) ? 1 : 0;
		$data{'CONTROL'}->{'FIELD_TYPE'} = "SABM" if ($control & 0xEF) == 0x2F;
		$data{'CONTROL'}->{'FIELD_TYPE'} = "DISC" if ($control & 0xEF) == 0x43;
		$data{'CONTROL'}->{'FIELD_TYPE'} = "DM" if ($control & 0xEF) == 0x0F;
		$data{'CONTROL'}->{'FIELD_TYPE'} = "UA" if ($control & 0xEF) == 0x63;
		$data{'CONTROL'}->{'FIELD_TYPE'} = "FRMR" if ($control & 0xEF) == 0x87;
		$data{'CONTROL'}->{'FIELD_TYPE'} = "UI" if ($control & 0xEF) == 0x03;
		$data{'CONTROL'}->{'FIELD_TYPE'} = "XID" if ($control & 0xEF) == 0xAF;
		$data{'CONTROL'}->{'FIELD_TYPE'} = "TEST" if ($control & 0xEF) == 0xE3;

		$data{'PID'} = sprintf("%02X", ord($frame[$location]));
		$location++;
		while ($location <= $#frame)
		{
			my $byte = $frame[$location];
			#my ($s, $ascii) = $translator->h2a(sprintf("%X",$byte));
			push @{$data{'INFO'}},  $byte;
			$location++;
		}
	}
	else
	{
		warn "Error: Couldn't determine the frame type.\n";
	}

	#Finally get the Frame Check Sequence which is the last two bytes of the frame
	#$data{'FCS'} = sprintf("%02X", ord($frame[$#frame - 1])) . sprintf("%02X", ord($frame[$#frame]));

	if (wantarray)
	{
		return %data;
	}
	else
	{
		return \%data;
	}
}

1;

__END__

=head1 SEE ALSO

 Device::TNC::KISS

AX.25 Link-Layer Protocol Specification L<http://www.tapr.org/pub_ax25.html>

=head1 AUTHOR

R Bernard Davison E<lt>bdavison@asri.org.auE<gt>

=head1 COPYRIGHT

Copyright (C) 2007, Australian Space Research Institute.

This module is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut
